client/rest/src/main/java/org/elasticsearch/client/Request.java

```diff
@@ -204,12 +204,37 @@ public boolean equals(Object obj) {
                 && endpoint.equals(other.endpoint)
                 && parameters.equals(other.parameters)
                 && Objects.equals(entity, other.entity)
-                && Arrays.equals(headers, other.headers)
+                && headers.equals(other.headers)
                 && httpAsyncResponseConsumerFactory.equals(other.httpAsyncResponseConsumerFactory);
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(method, endpoint, parameters, entity, Arrays.hashCode(headers), httpAsyncResponseConsumerFactory);
+        return Objects.hash(method, endpoint, parameters, entity, headers.hashCode(), httpAsyncResponseConsumerFactory);
+    }
+
+    static final class ReqHeader extends BasicHeader {
```

Comment:
Did you do it so we'd have a nice equals and hashcode? Maybe it'd be better to iterate by hand in those cases rather than make this. I'm not sure!

Reply:
++. Stick some javadoc on the class to explain that and merge away!